R/time_bins.R

```diff
@@ -94,161 +95,185 @@
+        stop("minimum `interval` value is less than available intervals")
       }
-
-      int_index <-
-        which(min_int <= df$max_ma & max_int >= df$min_ma)
-      df <- df[int_index,]
+      # Subset df
+      int_index <- which(df$max_ma >= min(interval) &
+                            df$min_ma <= max(interval))
     }
-  } else {
-    # Assign input scale to df
-    df <- scale
-    # Match column names to get_scale_data (or other user-input data)
-    col_indx <- match(c("name", "max_age", "min_age"), colnames(df))
-    # Column names
-    cnames <- c("interval_name", "max_ma", "min_ma")[which(!is.na(col_indx))]
-    # Remove NA matches
-    col_indx <- col_indx[!is.na(col_indx)]
-    if (length(col_indx) >= 1) {
-      colnames(df)[col_indx] <- cnames
+    # Sort int_index
+    int_index <- sort(int_index)
+    # Subset df for intervals
+    df <- df[int_index, ]
+    # Which rank should be used?
+    df <- df[which(df$rank == rank), ]
+    # Error handle
+    if(nrow(df) == 0) {
+      stop("No intervals are available for the defined interval range.")
     }
-
-    # Add mid_ma and duration myr
+  }
+  # Macrostrat dataframe scale -------------------------------------------
+  if (!scale %in% c("GTS2020", "GTS2012", "user")) {
+    # Try to get the time scale from Macrostrat
+    # Online and Macrostrat available?
+    tryCatch(
+      {
+        nslookup("macrostrat.org")
+      },
+      error = function(e) {
+        stop("Macrostrat is not available. Either the site is down or you are
+             not connected to the internet.",
+             call. = FALSE
+        )
+      }
+    )
+    url <- url(paste0("https://macrostrat.org/api/v2/defs/intervals",
+                      "?format=csv&timescale=",
+                      gsub(" ", "%20", scale)))
+    df <- tryCatch(
+      {
+        read.csv(url, header = TRUE, stringsAsFactors = FALSE)
+      },
+      error = function(e) {
+        stop("`name` does not match a built-in or Macrostrat time scale.",
+             call. = FALSE)
+      })
+    df <- df[, c("name", "b_age", "t_age", "abbrev", "color")]
+    colnames(df) <- c("interval_name", "max_ma", "min_ma", "abbr", "colour")
+    # Add mid_ma
     df$mid_ma <- (df$max_ma + df$min_ma) / 2
-    df$duration_myr <- df$max_ma - df$min_ma
-    # Rearrange columns
-    colnme <- colnames(df)
-    order_nme <- c("interval_name", "max_ma",
-                   "mid_ma", "min_ma", "duration_myr")
-    colnme <- colnme[!colnme %in% order_nme]
-    order_nme <- c(colnme, order_nme)
-    df <- df[, order_nme]
-
-    # Order by mid_ma
-    df <- df[order(df$mid_ma), ]
+    # Add duration
+    df$duration_myr <- (df$max_ma - df$min_ma)
+    # Add bin number
+    df$bin <- seq_len(nrow(df))
+    # Reorder dataframe
+    df <- df[, c("bin", "interval_name", "max_ma", "mid_ma",
+                 "min_ma", "duration_myr", "abbr", "colour")]
+    # Add colours based on https://stackoverflow.com/a/1855903/4660582
```

Comment:
I think we could do this for any df that has a "colour" column, including user-provided dfs?